### PR TITLE
Improve accessibility of cover images

### PR DESCRIPTION
**Pull request by xi**
I recently did an accessibility review of a local vufind installation. I thought that some of the changes could be useful for the wider community.

However, I don't want to set up a full development environment. So someone definitely has to check if these work in the general case.

### CODE REVIEW

File: themes/bootstrap3/templates/RecordDriver/DefaultRecord/cover.phtml
Comment (demiankatz) [inline themes/bootstrap3/templates/RecordDriver/DefaultRecord/cover.phtml]: Is the intent here to get rid of alt text on cover images because they are essentially decorative and do not add useful information for a non-sighted user? If so, I think that is probably a reasonable change.

What is the purpose of the tabindex change here?

File: themes/bootstrap3/templates/search/results.phtml
Comment (demiankatz) [inline themes/bootstrap3/templates/search/results.phtml]: I feel like the idea of turning the recommendation sidebar into an aside has been previously discussed and rejected, since the sidebar usually presents core functionality like faceting, which is more than (to quote the definition of aside) "only indirectly related to the document's main content." But maybe I'm misremembering. Maybe @crhallberg remembers this better than I do.

File: themes/bootstrap3/templates/layout/layout.phtml
Comment (demiankatz) [inline themes/bootstrap3/templates/layout/layout.phtml]: I can see the value in making this focusable for the benefit of keyboard users. However, I seem to recall @EreMaijala may have said something about the sr-only class in relation to the ongoing work to upgrade to Bootstrap 5, so I'm just flagging him here in case he has any concerns about this. (It's entirely possible that I'm misremembering something, and there's no problem here).

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks for sharing this, @xi -- I've left a few comments below. I defer to the expertise of others in areas of accessibility, but I'm leaving a few opinions and possible recollections of past related conversations. I'm hoping that @crhallberg and others can weigh in as well.

File: themes/bootstrap3/templates/RecordDriver/DefaultRecord/cover.phtml
Comment (xi) [reply themes/bootstrap3/templates/RecordDriver/DefaultRecord/cover.phtml]: As far as I am aware, the cover appears next to the title, which is also a link. So both the `alt` text and the link are redundant. The link is still useful for mouse users, but for keyboard users it is better to skip it.

File: themes/bootstrap3/templates/search/results.phtml
Comment (xi) [reply themes/bootstrap3/templates/search/results.phtml]: I think it is important to use *some* kind of [landmark](https://www.w3.org/TR/wai-aria-1.2/#landmark_roles) here.  Note that some of these would also require an accessible label. Maybe `region` is a better choice here? `form` could also be a good candidate, but I fear that this would have too much of an impact on the contents.

File: themes/bootstrap3/templates/layout/layout.phtml
Comment (xi) [reply themes/bootstrap3/templates/layout/layout.phtml]: In bootstrap 5 `.sr-only` and `.sr-only-focusable` have been renamed to `.visually-hidden` and `.visually-hidden-focusable`.

File: themes/bootstrap3/templates/RecordDriver/DefaultRecord/cover.phtml
Comment (crhallberg) [reply themes/bootstrap3/templates/RecordDriver/DefaultRecord/cover.phtml]: I don't think `tabindex=-1` is appropriate here, since we want the link to be reachable by tab. It may be redundant in the search results, but not on the record view.

File: themes/bootstrap3/templates/RecordDriver/DefaultRecord/cover.phtml
Comment (crhallberg) [inline themes/bootstrap3/templates/RecordDriver/DefaultRecord/cover.phtml]: If I understand correctly, links with no content, alt, or title will read out the href. I think an alt with the title is preferable to that, since our record IDs are usually numbers or UUIDs.

File: themes/bootstrap3/templates/layout/layout.phtml
Comment (crhallberg) [reply themes/bootstrap3/templates/layout/layout.phtml]: [In BS3, you need both `sr-only` and `sr-only-focusable`](https://getbootstrap.com/docs/3.4/css/#helper-classes-screen-readers) but this is an important change, thank you!

File: themes/bootstrap3/templates/search/results.phtml
Comment (crhallberg) [reply themes/bootstrap3/templates/search/results.phtml]: `region` will be more appropriate with the right aria-label. I could have sworn there was a `h2` denoting the search filters...

Comment (demiankatz) [reply themes/bootstrap3/templates/search/results.phtml]: There is no h2 specifically above the region where recommendations are placed; instead, each recommendation block within that region has its own h2. In most configurations, [the one in sidefacets.phtml](https://github.com/vufind-org/vufind/blob/dev/themes/bootstrap3/templates/Recommend/SideFacets.phtml#L36) ends up on top.

File: themes/bootstrap3/templates/RecordDriver/DefaultRecord/cover.phtml
Comment (xi) [reply themes/bootstrap3/templates/RecordDriver/DefaultRecord/cover.phtml]: I wasn't aware that the same code is also used on the record view, so thanks for pointing that out.

Just to be sure: I was under the impression that this was a link *to* the record view. Does the record view link to itself here?

Would it be possible to distinguish the two cases? If not I am fine with dropping this change,

Comment (demiankatz) [reply themes/bootstrap3/templates/RecordDriver/DefaultRecord/cover.phtml]: In search results, the thumbnail links to the record page. On the record page, the thumbnail links to a larger version of the image, when available. The thumbnail template is generic and receives the href it is supposed to link to from the calling code. We could add another parameter to the template if we need to differentiate behavior in a new way.

File: themes/bootstrap3/templates/RecordDriver/DefaultRecord/cover.phtml
Comment (xi) [reply themes/bootstrap3/templates/RecordDriver/DefaultRecord/cover.phtml]: To my knowledge, the `href` is never used as the accessible name. But you are correct in pointing out that [links need to have an accessible name](https://www.w3.org/TR/wai-aria-1.2/#link), so removing the alt text is in fact incorrect. I will revert that change.

**Comment by xi** [comment]
A general question: Would you prefer me to push new commits on top of this or amend the existing commits and force-push?

**Comment by demiankatz** [comment]
> A general question: Would you prefer me to push new commits on top of this or amend the existing commits and force-push?

Feel free to do whatever is easiest for you -- we generally "squash and merge" pull requests, so it's not necessary to keep a clean history here unless you want to keep separate changes in separate commits. (Keeping things separate is good for larger and more complex PRs, but I think this one is probably small enough to be understandable as a single commit).

File: themes/bootstrap3/templates/search/results.phtml
Comment (xi) [inline themes/bootstrap3/templates/search/results.phtml]: I added some new translatable strings. Should I also add these to the translation files?

Comment (demiankatz) [reply themes/bootstrap3/templates/search/results.phtml]: Yes, please. You only need to add them to en.ini and any other languages you are fluent in; that will be enough to ensure that they get translated into the other languages by our volunteers on the next translation cycle.

File: module/VuFind/src/VuFind/View/Helper/Root/Record.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/View/Helper/Root/Record.php]: The doc comment above needs to be updated to include this new parameter.

File: module/VuFind/src/VuFind/View/Helper/Root/Record.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/View/Helper/Root/Record.php]: Doc comment updates needed here too.

File: themes/bootstrap3/templates/RecordDriver/DefaultRecord/cover.phtml
Comment (demiankatz) [inline themes/bootstrap3/templates/RecordDriver/DefaultRecord/cover.phtml:15]: Might it still make sense to have an empty alt in situations where the image is not linked and/or the link is not unique, to reduce noise, since the cover images are essentially just decoration in these situations?

File: themes/bootstrap3/templates/RecordDriver/EDS/cover.phtml
Comment (demiankatz) [inline themes/bootstrap3/templates/RecordDriver/EDS/cover.phtml:2]: Is it a problem that the icon special case is no longer handled here? Or is this no longer relevant? (I confess that I haven't tested or investigated yet -- I just noticed that this was removed and didn't appear to have any equivalent in the revised code).

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks, @xi, this is looking very good. Just a few questions and comments below. If you're in a hurry to see some of these changes reflected in the dev branch right away, I don't mind cherry-picking the commits that fall outside of this review -- just let me know if that makes a difference to you or not.

File: themes/bootstrap3/templates/RecordDriver/DefaultRecord/cover.phtml
Comment (xi) [reply themes/bootstrap3/templates/RecordDriver/DefaultRecord/cover.phtml:15]: Every link needs to have an accessible name, so if there is a link there also needs to be an alt-text. But I can adapt the code for the case if there is no link.

File: themes/bootstrap3/templates/RecordDriver/EDS/cover.phtml
Comment (xi) [reply themes/bootstrap3/templates/RecordDriver/EDS/cover.phtml:2]: You are right, I missed that.

**Comment by xi** [comment]
> If you're in a hurry to see some of these changes reflected in the dev branch right away, I don't mind cherry-picking the commits that fall outside of this review -- just let me know if that makes a difference to you or not.

Thanks for asking! I am in no hurry. I know that sometimes it can be useful to avoid merge conflicts to cherry pick some commits early while others are still in discussion. Feel free to do that if you think this would be beneficial. But there is no pressure from my side.

File: themes/bootstrap3/templates/RecordDriver/EDS/cover.phtml
Comment (demiankatz) [reply themes/bootstrap3/templates/RecordDriver/EDS/cover.phtml:2]: I wonder if it would be slightly more clear to do something like:

```
$includeLinkInMarkup = $this->link && $cover !== false;
```

...and then use `$includeLinkInMarkup` in the if conditions on the open and close tags. This way, we don't have duplicated multi-part logic in multiple places, so we'll be less likely to get the code out of sync if we make future changes.

File: themes/bootstrap3/templates/RecordDriver/DefaultRecord/cover.phtml
Comment (demiankatz) [reply themes/bootstrap3/templates/RecordDriver/DefaultRecord/cover.phtml:15]: That's fair, but I wonder if the best thing to do would be to put `aria-hidden="true"` on the link for non-unique cover links. This way, our markup can be valid, but we can also hide useless repetitive "Cover Image" announcements from users who won't find them useful. What do you think?

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks for the progress, @xi. I think we're nearly done here, but a couple more small thoughts:

File: themes/bootstrap3/templates/RecordDriver/DefaultRecord/cover.phtml
Comment (xi) [reply themes/bootstrap3/templates/RecordDriver/DefaultRecord/cover.phtml:15]: My head is spinning a bit, so just to be sure:

- if no link: empty `alt`
- if non-unique link: non-empty `alt` and `aria-hidden`
- if unique link: non-empty `alt` and no `aria-hidden`

That's what I implemented.

I think it is a little weird that the "no link" case is a little less hidden than the "non-unique link" case. An alternative approach could be to always have a non-empty `alt` and add a `<div aria-hidden="true">` instead of the `<a>` for the "no link" case. What do you think?

Comment (demiankatz) [reply themes/bootstrap3/templates/RecordDriver/DefaultRecord/cover.phtml:15]: Yes, if that makes the template simpler and more readable, that makes sense to me and should have the same effect.

Comment (xi) [reply themes/bootstrap3/templates/RecordDriver/DefaultRecord/cover.phtml:15]: done

Comment (demiankatz) [reply themes/bootstrap3/templates/RecordDriver/DefaultRecord/cover.phtml:15]: One last question: since `div` is a block-level element and `a` is not, would it be better to use a `span` here for consistency of layout/flow?

Comment (xi) [reply themes/bootstrap3/templates/RecordDriver/DefaultRecord/cover.phtml:15]: That depends on the CSS. I assumed that since the `a` contains an `img` that there is some kind of special styling. In the EDS template it also contains `<div class="ajaxcover">`. But I don't know enough about the context to give a definitive answer. Do you think I should change it?

Note that `a` is no longer an inline element in HTML 5 https://stackoverflow.com/a/1828032.

Comment (demiankatz) [reply themes/bootstrap3/templates/RecordDriver/DefaultRecord/cover.phtml:15]: I'll defer to @crhallberg on this question, but it sounds like I'm probably just behind the times in my knowledge of the HTML standards. Thanks for clarifying that!

**Review by demiankatz — APPROVED** [review]
Thanks again, @xi -- this looks good to me now. However, I'm going to wait for more feedback from others before merging, just in case there are any considerations I am overlooking. Both @crhallberg and @EreMaijala are on vacation this week, but I hope to hear from one or both of them after they've returned and caught up.

**Comment by EreMaijala** [comment]
I think there may be one thing that complicates this: the record embedding support (see [List] section in searches.ini) in the results list. If enabled, the cover image is the only link to the full record page. Clicking the title just expands or collapses the record.

**Comment by demiankatz** [comment]
> I think there may be one thing that complicates this: the record embedding support (see [List] section in searches.ini) in the results list. If enabled, the cover image is the only link to the full record page. Clicking the title just expands or collapses the record.

Good catch, @EreMaijala! I guess maybe we need to use `$options->getListViewOption()` to decide on the appropriate behavior in the search results. Maybe when tabs or accordions are enabled, we do need to go back to the behavior where the image alt text is the title of the book. May take some thought to figure out how to do this in the least convoluted way -- there are a lot of code paths to consider here.

**Review by demiankatz — CHANGES_REQUESTED** [review]
Switching my review back to "request changes" until we can sort out the issue that @EreMaijala raised.

**Review by EreMaijala — CHANGES_REQUESTED** [review]
Toggling my review status as well until the issue is resolved.

**Comment by demiankatz** [comment]
Thanks, @EreMaijala. Since this process is taking a while, I went ahead and cherry-picked the two uncontroversial commits out of this PR into dev and rebased. Might as well benefit from those pieces while we work through the challenges with this last part.

**Comment by demiankatz** [comment]
This hasn't moved in a little while, so I'm taking the 10.0 milestone off of it for the moment -- I'm not sure if it will get done in time for that release. I have also retitled the PR to reflect the parts that remain, since the other accessibility items were merged separately.

@xi, please let us know if you need more help from our end to continue moving this forward.

**Comment by xi** [comment]
Oh, sorry, it seems like there was a misunderstanding. I was actually waiting for you to tell me what to do. I know next to nothing about the issue that @EreMaijala pointed out, so I cannot decide how to deal with it.

**Comment by demiankatz** [comment]
@xi, I confess that this is a bit hard for me to wrap my head around as well, but maybe as a starting point we need to be a little more explicit about the behaviors. Maybe rather than using the $linkUnique argument to the getCover methods, we should add two different arguments:

$useTitleAsAltText = should we use the title of the item as the image alt text (true) or the generic "Cover Image" text (false)?

$ariaHidden = should we hide this element from screen readers and keyboard access (true) or not (false)?

The tricky part is figuring out how to set these arguments based on circumstances.

You can see the behavior that @EreMaijala mentions by setting "view" to either "tabs" or "accordion" in the [List] section of searches.ini. As noted earlier, it should be possible to use the Search Options object to figure out when it is turned on. Unfortunately, my memory of exactly how this is implemented is a bit hazy (I didn't write it, and it's been years since I reviewed it) and I'm short on time to refresh my memory... but that may be necessary if you need more help. Please let me know what you think! I probably won't have the capacity to do anything with this until after 10.0 is released due to the many tasks I have to take care of leading up to the release, but if it can wait that long, I don't mind helping further when I'm able... and maybe others can step up in the meantime. :-)

**Comment by xi** [comment]
Just to be sure: Why do we want to use the title as alt text when record embedding is used? @EreMaijala wrote:

> If enabled, the cover image is the only link to the full record page. Clicking the title just expands or collapses the record.

For me that seems to imply `linkUnique = true`, but not that the text of the link must be the record title. Maybe  "open full page" would be even better in that case.

**Comment by demiankatz** [comment]
> Just to be sure: Why do we want to use the title as alt text when record embedding is used? @EreMaijala wrote:
> 
> > If enabled, the cover image is the only link to the full record page. Clicking the title just expands or collapses the record.
> 
> For me that seems to imply `linkUnique = true`, but not that the text of the link must be the record title. Maybe "open full page" would be even better in that case.

I think the issue in this case is that every search result will have its own image link that opens the record page -- so there presumably needs to be a way to differentiate them from one another.

**Comment by xi** [comment]
Unfortunately I don't think I have time to work on this anymore. Thanks for all the constructive feedback! I hope that someone will be able to pick this up.

**Comment by demiankatz** [comment]
Thanks, @xi; I'm glad we were at least able to get all of your other improvements merged, and I'm sure we'll find time for a community discussion about how to finish this work at some point -- though likely not until after the 10.0 release is completed.

**Comment by crhallberg** [comment]
Quite late to this and I'm not sure if it simplifies it at this point, but if an image is the only child of a link, the image needs an alt attribute or the link needs an attribute. We can put an sr-only title inside the link allowing us to put alt="" on the image, since these are technically decorative.

https://russmaxdesign.github.io/accessible-forms/accessible-name-link03.html
image

**Comment by demiankatz** [comment]
@crhallberg, I think the complication here is that the thumbnails are _usually_ decorative, but there are some edge cases where they are not -- for example, if you turn on record embedding (tabs/accordions), then the thumbnail is the only way to get to the actual record page. Record embedding is problematic for other reasons, but we need to either account for that problem in our solution here or else change the way embedding behaves.

**Comment by EreMaijala** [comment]
@demiankatz I think we need to change images to be purely decorative even with embedding. This of course means that we need "Show full record" link or something like that, which pollutes the display, but the current mechanism isn't obvious or easily discoverable. That'd get rid of this issue, and would e.g. make it possible to display a large image popup on click instead of going to the record page.

**Comment by demiankatz** [comment]
> @demiankatz I think we need to change images to be purely decorative even with embedding. This of course means that we need "Show full record" link or something like that, which pollutes the display, but the current mechanism isn't obvious or easily discoverable. That'd get rid of this issue, and would e.g. make it possible to display a large image popup on click instead of going to the record page.

That seems like a reasonable solution, and we already have a "View Full Record" translation (currently used for linking to record URLs from results emails, so with a matching semantic meaning) that could be easily used for this. Maybe we should add that in a separate PR and then circle back to this one. Does anyone have time to do that? If not, I'll try to put it on my to-do list somewhere.

**Comment by xi** [comment]
I fixed the merge conflicts, so hopefully this can go into 10.1.

**Review by demiankatz — CHANGES_REQUESTED** [review]
@xi, it looks like the templates are out of sync between the themes. You can fix this with: `vendor/bin/phing update-bootstrap5`. Do you mind doing that and committing the change? If you don't have time, I'll do it when I get home next week...

**Comment by xi** [comment]
> it looks like the templates are out of sync between the themes. You can fix this with: `vendor/bin/phing update-bootstrap5`

It would be great if you could do that. I don't really have the relevant development setup available right now.

**Comment by demiankatz** [comment]
@xi, I had a few moments, so I just did it. I may not have time to fully re-test/re-review until I'm home next week, though. Stay tuned, and thanks for your help! :-)

File: themes/bootstrap3/templates/RecordDriver/DefaultRecord/cover.phtml
Comment (demiankatz) [inline themes/bootstrap3/templates/RecordDriver/DefaultRecord/cover.phtml]: Since this PR was originally created, the behavior of the cover images in the search results has changed; instead of linking to the record page, it now pops up a larger version of the image just like on the record page. Now that the behavior is consistent everywhere, do we no longer need this special case?

**Review by demiankatz — CHANGES_REQUESTED** [review]
@xi, I've done a bit of testing and found that things weren't quite working. I refactored the code to solve the problem, but I have one last question that I think we should decide about before finalizing this....

File: themes/bootstrap3/templates/RecordDriver/DefaultRecord/cover.phtml
Comment (xi) [reply themes/bootstrap3/templates/RecordDriver/DefaultRecord/cover.phtml]: In that case I don't think we need this special case. There might even be many more changes we don't need anymore.

When I originally opened this pull request there was 0af81e1ba02f09ee0b4122581af20dec236d283f to fix redundant links and 1a93ec341c8de803912cf4e719637a106aea286b to fix redundant alternative texts. If the links are no longer redundant, the only thing we really need to change are the alt texts.

Comment (demiankatz) [reply themes/bootstrap3/templates/RecordDriver/DefaultRecord/cover.phtml]: Sounds like that could greatly simplify everything. I think if we can eliminate the linkUnique flag that will make the changes here (and the code in general) a bit less potentially confusing.

I don't mind taking a stab at removing $linkUnique if that would be helpful. Is the idea then that we'll leave the aria-hidden div in place for non-linked cover images since those are not useful/meaningful, but we'll leave linked cover images unhidden with the normal tab order, but with appropriate alt text?

Comment (xi) [reply themes/bootstrap3/templates/RecordDriver/DefaultRecord/cover.phtml]: I recommend doing this from scratch, just changing the alt texts.

Comment (demiankatz) [reply themes/bootstrap3/templates/RecordDriver/DefaultRecord/cover.phtml]: There are a couple other bits of refactoring here that I think would be useful to keep... but it shouldn't be hard to reset most of the changes. I'll adjust shortly!

Comment (demiankatz) [reply themes/bootstrap3/templates/RecordDriver/DefaultRecord/cover.phtml]: Okay, I have rebased and simplified the code -- what do you think of this version?

**Review by demiankatz — APPROVED** [review]
All tests are passing and this looks reasonable to me, so I'm going to give this my approval. However, I'll wait to hear from @xi and @EreMaijala before actually merging, in case any further adjustments are needed.

File: themes/bootstrap3/templates/RecordDriver/DefaultRecord/cover.phtml
Comment (xi) [reply themes/bootstrap3/templates/RecordDriver/DefaultRecord/cover.phtml]: I must admit my head is spinning from all the revisions this have been going through. The code looks reasonable to me, but the devil is in the details.

Comment (demiankatz) [reply themes/bootstrap3/templates/RecordDriver/DefaultRecord/cover.phtml]: Agreed, it has been a long and twisting road to get this far. But I think the bottom line is that we're using uniform alt text on the cover images now because they're no longer being used as record links, and we're aria-hiding the thumbnails when they don't serve as links since they're purely decorative in that instance... and those seem like reasonable changes that should be safe to commit. If there are additional adjustments needed, we can always follow up later!

**Review by EreMaijala — APPROVED** [review]

**Comment by demiankatz** [comment]
I'll go ahead and merge this now. Feel free to open a follow-up if you have any further ideas, @xi, and thanks for getting this started!

**Pull request merged**
